design/debug-collector.md

```diff
@@ -0,0 +1,68 @@
+
+# Motivation
+
+The user may wish to collect ekotrace logs from a microcontroller with no networking capabilities. This tool allows for collection using JTAG/SWD wire protocols instead of UDP.
+
+# Technical Strategy
+
+## Memory Access
+
+To access memory correctly, the program needs to know the word size and endianness of the target. The user can either pass an ELF file containing this information into the CLI: `--elf`, or they can specify the information explicitly: `--word-size 32 --big-endian`.
+
+### Choice of backend
+
+The collector will access memory on the target device either by using the probe-rs Rust library, or by talking to a gdb server that is attached to the target. This choice will be given to the user of the CLI, using `--attach` for probe-rs, or `--gdb-host` to use gdb.
+
+#### probe-rs
+
+probe-rs provides functions to attach to chips over a usb connection, as well as to read from or write to memory locations.
+
+To use this option, users must specify the type of chip that they wish to connect to, which is required by probe-rs: `--attach stm32`.
+
+#### gdb
+
+If the user already has a GDB server that attaches to the chip set up (e.g. OpenOCD), or would rather use such a tool to attach to the target, then they can specify the address of the gdb server that they are running: `--gbd-host localhost:3333`.
+
+In this case, the debug-collector will communicate to the given gdb server over TCP in order to read and write to memory.
+
+### repr(C)
+
+Rust's ABI is undefined, so the representation of structures in memory is unpredictable across compiler versions and architectures. To allow the debug-collector to understand the raw binary it reads from the chip, the ekotrace log data structures must have the `repr(C)` attribute, which tells the compiler to use C's representation of structs in binary. To find the fields of `repr(C)` structs, the debug-collector uses constant offsets for each given word size. This method is fragile to implementation changes to the structs that the debug-collector is accessing.
+
+We would have to add this attribute to `DynamicHistory` and `ChunkedReportState` in ekotrace as well as to `FixedSliceVec` in fixed-slice-vec.
+
+## Symbol resolution
+
+For each tracer, the user will either pass in the memory location of a pointer to the tracer, or will pass in the name of the symbol of the pointer to the tracer. If the latter is used, then the user would also need to pass in an ELF file containing the symbol table of the program running on the target device. The debug-collector will automatically find the memory location of the given symbol by looping through the given symbol table.
+
+## Locking
```

Comment:
This can get tricky in a `no_std` context.
Core does have [AtomicBool](https://doc.rust-lang.org/core/sync/atomic/struct.AtomicBool.html), but as noted in the [Portability Section](https://doc.rust-lang.org/core/sync/atomic/index.html#portability), atomic types are not available on all platforms.
I think often implementations fall back to using `UnsafeCell<T>`, and use `intrinsics::atomic_*`. We might find ourselves doing something similar if we intend on supporting those oddball platforms.
I think for now, saying we'll support platforms that have atomics is probably ok.